core/src/main/java/org/elasticsearch/indices/IndicesService.java

```diff
@@ -633,20 +643,22 @@ public void deleteShardStore(String reason, ShardId shardId, ClusterState cluste
     }
 
     /**
-     * This method returns true if the current node is allowed to delete the
-     * given index. If the index uses a shared filesystem this method always
-     * returns false.
+     * This method returns true if the current node is allowed to delete the given index.
+     * This is the case if the index is deleted in the metadata or there is no allocation
+     * on the local node and the index isn't on a shared file system.
      * @param index {@code Index} to check whether deletion is allowed
      * @param indexSettings {@code IndexSettings} for the given index
+     * @param indexIsDeleted true if the index has been deleted from the cluster,
+     *                       but not yet removed from this node due to allocation on it
      * @return true if the index can be deleted on this node
      */
-    public boolean canDeleteIndexContents(Index index, IndexSettings indexSettings, boolean closed) {
-        final IndexService indexService = indexService(index);
-        // Closed indices may be deleted, even if they are on a shared
-        // filesystem. Since it is closed we aren't deleting it for relocation
-        if (indexSettings.isOnSharedFilesystem() == false || closed) {
+    public boolean canDeleteIndexContents(Index index, IndexSettings indexSettings, boolean indexIsDeleted) {
+        // index contents can be deleted if the index is not on a shared file system,
+        // or if its on a shared file system but its a deleted index
+        if (indexIsDeleted || indexSettings.isOnSharedFilesystem() == false) {
+            final IndexService indexService = indexService(index);
             if (indexService == null && nodeEnv.hasNodeFile()) {
-                return true;
+                return nodeEnv.indexDirectoryExists(index, indexSettings);
```

Comment:
thinking about this some  more - do we need to change this at all? doesn't seem like #17265 has this change